toponetx/utils/normalization.py

```diff
@@ -0,0 +1,383 @@
+    """
+    assert L.shape[0] == L.shape[1]
+    topeig = spl.eigsh(L.asfptype(), k=1, which="LM", return_eigenvectors=False)[0]
+    return Lx * (1.0 / topeig)
+
+
+def compute_kipf_adjacency_normalized_matrix(
+    A_opt, add_identity=False, identity_multiplier=1.0
+):
+    """
+    Normalize the adjacency matrix using Kipf's normalization.
+
+    Typically used to normalize adjacency matrices.
+
+    Parameters
+    ----------
+    A_opt : numpy.ndarray
+        The adjacency matrix.
+    add_identity : bool, optional
+        Determines if the identity matrix is to be added to the adjacency matrix.
+    identity_multiplier : float, optional
+        A multiplier of the identity. This parameter is helpful for higher order
+        (co)adjacency matrices where the neighbor of a cell is obtained from multiple sources.
+
+    Returns
+    -------
+    scipy.sparse.csr_matrix
+        The normalized adjacency matrix.
+
+    Notes
+    -----
+    This normalization is based on Kipf's formulation, which computes the row-sums,
+    constructs a diagonal matrix D^(-0.5), and applies the normalization as D^(-0.5) * A_opt * D^(-0.5).
+    """
+    if add_identity:
+        size = A_opt.shape[0]
+        eye = diags(size * [identity_multiplier])
+        A_opt = np.abs(A_opt) + eye
+    else:
+        A_opt = np.abs(A_opt)
+    rowsum = np.array(A_opt.sum(1))
+    r_inv_sqrt = np.power(rowsum, -0.5).flatten()
+    r_inv_sqrt[np.isinf(r_inv_sqrt)] = 0.0
+    r_mat_inv_sqrt = diags(r_inv_sqrt)
+    A_opt_normalized = A_opt.dot(r_mat_inv_sqrt).transpose().dot(r_mat_inv_sqrt)
+
+    return A_opt_normalized
+
+
+def compute_xu_asymmetric_normalized_matrix(B, is_sparse=True):
+    """
+    Compute Xu's normalized asymmetric matrix.
+
+    Typically used to normalize boundary operators.
+
+    Parameters
+    ----------
+    B : numpy.ndarray or scipy.sparse.csr_matrix
+        The asymmetric matrix.
+    is_sparse : bool, optional
+        If True, treat B as a sparse matrix.
+
+    Returns
+    -------
+    numpy.ndarray or scipy.sparse.csr_matrix
+        The normalized asymmetric matrix.
+
+    Notes
+    -----
+    This normalization is based on Xu's formulation, which computes the diagonal matrix D^(-1)
+    and multiplies it with the asymmetric matrix B.
+    """
+    if is_sparse:
+        rowsum = np.array(np.abs(B).sum(1))
+        r_inv = np.power(rowsum, -1).flatten()
+        r_inv[np.isinf(r_inv)] = 0.0
+        r_mat_inv = diags(r_inv)
+        normalized_matrix = r_mat_inv.dot(B)
+        return normalized_matrix
+    else:
+        Di = np.sum(np.abs(B), axis=1)
+        Di2 = np.diag(np.power(Di, -1))
+        Di2[np.isinf(Di2)] = 0.0
+        B = np.array(B)
+        normalized_matrix = Di2.dot(B)
+        return normalized_matrix
+
+
+def compute_bunch_normalized_matrices(B1, B2):
+    """Get Bunch normalization.
+
+    Parameters
+    ----------
+    B1 : numpy array or scipy csr_matrix
```

Comment:
`numpy array` and `scipy csr_matrix` should be replaced by `np.ndarray` and `scipy.sparse.csr_matrix`, respectively, throughout the document